runtime/kusama/src/polkadot_messages.rs

```diff
@@ -0,0 +1,620 @@
+				.saturating_sub(if include_pay_dispatch_fee_cost {
+					0
+				} else {
+					bp_polkadot::PAY_INBOUND_DISPATCH_FEE_WEIGHT
+				})
+				.saturating_add(message_dispatch_weight),
+			size: message_payload_len
+				.saturating_add(bp_kusama::EXTRA_STORAGE_PROOF_SIZE)
+				.saturating_add(bp_polkadot::TX_EXTRA_BYTES),
+		}
+	}
+
+	fn transaction_payment(transaction: MessageTransaction<Weight>) -> bp_polkadot::Balance {
+		// we don't have a direct access to the value of multiplier of Polkadot chain
+		// => it is a messages module parameter
+		let multiplier = PolkadotFeeMultiplier::get();
+		let per_byte_fee = bp_polkadot::TRANSACTION_BYTE_FEE;
+		transaction_payment(
+			bp_polkadot::BlockWeights::get().get(DispatchClass::Normal).base_extrinsic,
+			per_byte_fee,
+			multiplier,
+			|weight| bp_polkadot::WeightToFee::calc(&weight),
+			transaction,
+		)
+	}
+}
+
+impl TargetHeaderChain<ToPolkadotMessagePayload, bp_polkadot::AccountId> for Polkadot {
+	type Error = &'static str;
+	type MessagesDeliveryProof =
+		messages_source::FromBridgedChainMessagesDeliveryProof<bp_polkadot::Hash>;
+
+	fn verify_message(payload: &ToPolkadotMessagePayload) -> Result<(), Self::Error> {
+		messages_source::verify_chain_message::<WithPolkadotMessageBridge>(payload)
+	}
+
+	fn verify_messages_delivery_proof(
+		proof: Self::MessagesDeliveryProof,
+	) -> Result<(LaneId, InboundLaneData<bp_kusama::AccountId>), Self::Error> {
+		messages_source::verify_messages_delivery_proof::<
+			WithPolkadotMessageBridge,
+			Runtime,
+			crate::PolkadotGrandpaInstance,
+		>(proof)
+	}
+}
+
+impl SourceHeaderChain<bp_polkadot::Balance> for Polkadot {
+	type Error = &'static str;
+	type MessagesProof = messages_target::FromBridgedChainMessagesProof<bp_polkadot::Hash>;
+
+	fn verify_messages_proof(
+		proof: Self::MessagesProof,
+		messages_count: u32,
+	) -> Result<ProvedMessages<Message<bp_polkadot::Balance>>, Self::Error> {
+		messages_target::verify_messages_proof::<
+			WithPolkadotMessageBridge,
+			Runtime,
+			crate::PolkadotGrandpaInstance,
+		>(proof, messages_count)
+		.and_then(verify_inbound_messages_lane)
+	}
+}
+
+/// Verify that lanes of inbound messages are enabled.
+fn verify_inbound_messages_lane(
+	messages: ProvedMessages<Message<bp_polkadot::Balance>>,
+) -> Result<ProvedMessages<Message<bp_polkadot::Balance>>, &'static str> {
+	let allowed_incoming_lanes = [[0, 0, 0, 0]];
+	if messages.keys().any(|lane_id| !allowed_incoming_lanes.contains(lane_id)) {
+		return Err(INBOUND_LANE_DISABLED)
+	}
+	Ok(messages)
+}
+
+impl SenderOrigin<AccountId> for Origin {
+	fn linked_account(&self) -> Option<AccountId> {
+		match self.caller {
+			// in benchmarks we accept messages from regular users
+			#[cfg(feature = "runtime-benchmarks")]
+			crate::OriginCaller::system(frame_system::RawOrigin::Signed(ref submitter)) =>
+				Some(submitter.clone()),
+
+			OriginCaller::Council(_) => AllowedMessageSender::get(),
```

Comment:
That's another questionable part, which I've mentioned in description. It means that we'll only accept messages with `OriginCaller::Council` origin